glTF-Toolkit/src/GLTFSpecularGlossinessUtils.cpp

```diff
@@ -0,0 +1,264 @@
+        }
+        catch (GLTFException)
+        {
+            throw GLTFException("Failed to load specular glossiness texture.");
+        }
+    }
+
+    ScratchImage metallicRoughnessTexture;
+    ScratchImage modulatedDiffuseTexture;
+
+    ConvertTextureSpecularGlossinessToMetallicRoughness(
+        metallicRoughnessTexture,
+        modulatedDiffuseTexture,
+        diffuseTexture,
+        diffuseFactorIn, // will be baked into texture
+        specularGlossinessTexture,
+        specularFactor);
+
+    Material::PBRMetallicRoughness gltfPBRMetallicRoughness;
+
+    {
+        DirectX::ScratchImage converted;
+        if (FAILED(DirectX::Convert(*metallicRoughnessTexture.GetImage(0, 0, 0), DXGI_FORMAT_B8G8R8X8_UNORM, DirectX::TEX_FILTER_SRGB_IN, DirectX::TEX_THRESHOLD_DEFAULT, converted)))
+        {
+            throw GLTFException("Failed to convert texture to DXGI_FORMAT_B8G8R8X8_UNORM for processing.");
+        }
+
+        auto metallicRoughnessPath = GLTFTextureUtils::SaveAsPng(&converted, "metallicRoughness_" + material.id + ".png", outputDirectory);
+        auto metallicRoughnessImageId = GLTFTextureUtils::AddImageToDocument(resultDoc, metallicRoughnessPath);
+        Texture mrTexture;
+        mrTexture.samplerId = samplerId;
+        mrTexture.imageId = metallicRoughnessImageId;
+        gltfPBRMetallicRoughness.metallicRoughnessTexture.textureId = resultDoc.textures.Append(mrTexture, AppendIdPolicy::GenerateOnEmpty).id;
+    }
+
+    {
+        DirectX::ScratchImage converted;
+        if (FAILED(DirectX::Convert(*modulatedDiffuseTexture.GetImage(0, 0, 0), DXGI_FORMAT_B8G8R8A8_UNORM_SRGB, DirectX::TEX_FILTER_DEFAULT, DirectX::TEX_THRESHOLD_DEFAULT, converted)))
+        {
+            throw GLTFException("Failed to convert texture to DXGI_FORMAT_B8G8R8A8_UNORM_SRGB for processing.");
+        }
+        auto diffusePath = GLTFTextureUtils::SaveAsPng(&converted, "diffuse_" + material.id + ".png", outputDirectory, &GUID_WICPixelFormat32bppBGRA);
+        auto diffuseImageId = GLTFTextureUtils::AddImageToDocument(resultDoc, diffusePath);
+        Texture diffusGltfTexture;
+        diffusGltfTexture.samplerId = samplerId;
+        diffusGltfTexture.imageId = diffuseImageId;
+        gltfPBRMetallicRoughness.baseColorTexture.textureId = resultDoc.textures.Append(diffusGltfTexture, AppendIdPolicy::GenerateOnEmpty).id;
+    }
+
+    resultMaterial.metallicRoughness = gltfPBRMetallicRoughness;
+    resultDoc.materials.Replace(resultMaterial);
+
+    return resultDoc;
+}
+
+
+Document GLTFSpecularGlossinessUtils::ConvertMaterials(std::shared_ptr<IStreamReader> streamReader, const Document & doc, const std::string & outputDirectory)
+{
+    Document resultDocument(doc);
+    for (const auto& material : doc.materials.Elements())
+    {
+        resultDocument = ConvertMaterial(streamReader, resultDocument, material, outputDirectory);
+    }
+
+    resultDocument.extensionsUsed.erase(KHR::Materials::PBRSPECULARGLOSSINESS_NAME);
```

Comment:
does this fail if the extension is not there?

Reply:
No idempotent in that case.